### DISCUSSION TITLE
Boosters Overhaul

### DISCUSSION THREAD

**Discussion by spacek531** [OpenRCT2/OpenRCT2#23119] [Feature requests]
I realize I never made a document outlining everything I want to do regarding boosters, so here it is.

There are several components to the overhaul:

## Legacy Behavior Flag

This feature has an open PR https://github.com/OpenRCT2/OpenRCT2/pull/25920

Because of the specifics of this change, a flag will be implemented to preserve backwards-compatibility.  This overhaul opens up a lot of new features that OpenRCT2 users did not account for, so some hacks will break if backwards compatibility is not explicitly added. It is unfortunate to keep code around specifically for deprecated behavior, but necessary to improve OpenRCT2. To prevent flag bloat, the amount of changes that are implemented at one time should be maximized.

Currently, the RTD has a LegacyBoosterSettings and BoosterSettings fields. The BoosterSettings field is currently unused - it will be used when the overhaul is complete.

## Unify Booster Speeds

This feature is fully implemented.

Currently, all track speed values use 4 bits for a total of 15 nonzero values. Giga and Junior coaster boosters have double or half the range by multiplying or dividing the values by 2 (the track's 4 bit speed value is multiplied by 2 beforehand so there is no loss of precision). This is fine in a vacuum, but RCT has an unintuitive simulation that uses the vehicle's RTD to determine the multiply/divide choice instead of the track's RTD. When hacking rides, it is common to use a non-giga coaster ride type but desire giga coaster booster speeds. Even if the ride vehicle passes over a giga coaster booster (via merging) set to 215 km/h, a non-giga coaster ride will go 67 km/h instead. 

![image](https://github.com/user-attachments/assets/a3ede27c-e192-4e7b-a407-b59884bd6f59)

The overhaul will address this by increasing the 4 bits of BrakeBoosterSpeed to a minimum of 6 bits and maximum of 8 bits such that the new speed regime reaches as high as the giga coaster speed, at the granularity of junior coaster. On import, all track elements will have their speeds converted to the unified speed values, and vehicles will not use BoosterSpeedFactor to convert the BrakeBoosterSpeed value to its final value. Backwards compatibility will be maintained using a legacy behavior flag, which will force a conversion that mimics the legacy behavior (currently implemented). TD6 serialization is a small can of worms the size of a large can of worms, but it mostly boils down to giving the user an alert if the speed cannot be properly serialized, and changing the message depending on exactly what could not be properly serialized (out of range/too precise). TD6 deserialization is not an issue.

A UI bug that displays speeds lower than actual value when building giga coaster boosters will be fixed.

## Booster Power From Track

This feature is fully implemented.

Booster strength is determined the same way as speed regime. A looping coaster, which has weak boosters, will only ever get looping coaster booster power when a train passes over a booster, regardless of which ride the booster belongs to. Likewise, there is the LSM behavior of the Reverse Freefall Coaster which would be nice as a ride design option for other ride types.

The overhaul will address this by using the booster power of the track's RTD instead of the vehicle's RTD. When the legacy flag is set, it will use the vehicle's RTD instead of the track's RTD.

With this change, OpenRCT2 is set up for a future update that allows booster pieces to set their strength value individually which is the end goal of booster power.

## Booster Modes

This feature is in the planning stages.

Currently, boosters boost a train going forwards and do nothing when the train is going backwards. Creating swing launches requires a lot of planning to work around this limitation. Adding up to 7 operating modes will help users create rides they want to.

- Boosting forward, coasting reverse (BrakeBoosterSpeed > 0, mode 0)
- Boosting reverse, coasting forward (BrakeBoosterSpeed < 0, mode 0)
- Boosting forward, passive braking reverse (BrakeBoosterSpeed > 0, mode 1)
- Boosting reverse, passive braking forward (BrakeBoosterSpeed < 0, mode 1)
- Boosting forward, active braking in reverse (BrakeBoosterSpeed > 0, mode 2)
- Boosting reverse, active braking in forward (BrakeBoosterSpeed < 0, mode 2)
- Boosting forward, boosting reverse (mode 3)

The sticking point is that the free bits in the TrackElement struct are not ideal. The bits I would like to use for this feature are Flags2 bits 4 and 5 (TRACK_ELEMENT_FLAGS2_HAS_GREEN_LIGHT, TRACK_ELEMENT_FLAGS2_BRAKE_CLOSED) for booster mode and Flags2 bit 7 (unused) for direction. Powered vehicles use a sign-magnitude implementation for its powered speed, where the magnitude bits span an entire byte and the sign bit is in the flags. I would like to use this for track elements as well, to unify things between track and vehicle.

This doesn't require backwards compatibility as far as I know.

## Block brakes clamp speed when departing

Coasters are too slow coming off block brakes https://github.com/OpenRCT2/OpenRCT2/discussions/22117

## Allowing boosters on hacked rides

Rides that don't get boosters cannot use boosters even when they are hacked onto the ride https://github.com/OpenRCT2/OpenRCT2/discussions/21752

This is partially addressed by merging, but it would be nice if merges were not required. The split between LegacyBoosterSettings and BoosterSettings is for this feature.

## Set BrakeBoosterSpeed from Tile Inspector

Spinner measuring internal units, possibly with the localized speed adjacent.

**Comment by CometWolf2323** [comment]
I cannot wait for it to look so amazing for that redesign for Boosters
